pm: device power management improvements

gmarull opened the pull request:
Summary:

- Remove usage of local states, the device PM subsystem already holds device PM state
- The "ctrl_command" is removed, since the PM control callback will now always be called to set the state, not to retrieve it.
- The state is passed as a value, since it does not need to be written by the PM control callback
- The callback is only called for a state change, thus simplifying driver operation (no need for state change checks)
- Converged _low power_ states to `PM_DEVICE_STATE_SUSPEND`. This makes interoperability between system and device PM easier. The low power state is still kept, but its future is unclear and needs some discussion.
- Device PM control function return codes are now documented.
- Multiple implementation cleanups.
- Introduction of the _actions_ concept. Devices will now receive actions instead of states, for example, a device will be told to "suspend" or to "resume".

gmarull committed:
pm: device: remove usage of local states

The device PM subsystem already holds the device state, so there is no
need to keep duplicates inside the device. The pm_device_state_get has
been refactored to just return the device state. Note that this is still
not safe, but the same applied to the previous implementation. This
problem will be addressed later.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove usage of ctrl_command

The callback is now invoked to set the device PM state in all cases, so
the usage of ctrl_command is redundant.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove ctrl_command callback argument

The ctrl_command is not used anymore, so remove it from the callback
signature.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
doc: pm: update documentation

Update documentation so that latest changes are reflected.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: remove pointer usage for state

Since the state is no longer modified by the device PM callback, just
use the state value.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: improve docs and arg names for state set

Improve the docstrings of the pm_device_state_set function and change to
shorter argument variable names.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: do not call device if already on the given state

If the device is already at the given state, do not call the device PM
control function. This makes sure that devices are only called to change
from one state to another.

Even though asynchronous device PM is completely broken, transitional
states are considered too.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: simplify state change check logic

The device PM control function will only be called if the requested
state is different from the current one. A significant amount of drivers
were checking for state changes, now unnecessary. This patch removes all
this redundant logic.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: rename PM_DEVICE_STATE_SUSPEND to PM_DEVICE_STATE_SUSPENDED

The verb tense for the suspended state was not consistent with other
states. The likely reason: state was being used as a command/action.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: cleanup pm control callback implementations

- Return -ENOTSUP if the requested state is not supported
- Remove redundant "noop style" functions.
- Use switch everywhere to handle requested state (not necessary in all
  drivers, but better take off with consistency in place after current
  changes).

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: fix incorrect usages of PM_DEVICE_STATE_OFF

According to the documentation the OFF state has to be used when the
devices is fully turned off, ie, power removed. Most drivers were using
a sort of fall-through for all non-active states, leading to behaviors
not following the specifications.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: add type for device PM control callback

Add a type and documentation for the device PM control callback. This
way possible return codes and its meaning are documented.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: use actions for device PM control

Instead of passing target states, use actions for device PM control.
Actions represent better the meaning of the callback argument.
Furthermore, they are more future proof as they can be suitable for
other PM actions that have no direct mapping to a state. If we compare
with Linux, we could have a multi-stage suspend/resume. Such scenario
would not have a good mapping when using target states.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

carlescufi reviewed (approved):
+1 to the approach after discussion with @gmarull

ceolin commented on drivers/interrupt_controller/intc_ioapic.c:
```diff
@@ -314,35 +310,21 @@ int ioapic_resume_from_suspend(const struct device *port)
+		ret = ioapic_suspend(dev);
+		break;
+	default:
+		ret = -ENOTSUP;
```
Original code is handling `PM_DEVICE_STATE_LOW_POWER`. Here you are not doing it and it will return `-ENOTSUP` instead of `0`.

gmarull replied:
It's not handling it really, it does nothing. 
```
		case PM_DEVICE_STATE_LOW_POWER:
			break;
```

I can leave as it is and fix this problem later. I don't think PM subsys is handling `-ENOTSUP` correctly as of today.

gmarull replied:
added the case back, will fix these sort of issues later.

ceolin commented on drivers/spi/spi_nrfx_spi.c:
```diff
@@ -276,51 +273,32 @@ static int init_spi(const struct device *dev)
+	case PM_DEVICE_STATE_LOW_POWER:
+	case PM_DEVICE_STATE_SUSPEND:
+	case PM_DEVICE_STATE_OFF:
+		nrfx_spi_uninit(&config->spi);
```
you are unconditionally calling `nrfx_spi_uninit` while the original code just call it if the current state is `ACTIVE`.

gmarull replied:
this is precisely the purpose of this PR: remove this redundant logic. There is no need to do this check, since the device will be in active state if it reaches this point.

ceolin replied:
One calls `pm_device_state_set(dev, PM_DEVICE_STATE_LOW_POWER);` if the device is ACTIVE it will result in this code calling `nrfx_spi_uninit(&config->spi);`. Now the device is `PM_DEVICE_STATE_LOW_POWER`.

Now, if we do: `pm_device_state_set(dev, PM_DEVICE_STATE_OFF);` this function will end up calling  `nrfx_spim_uninit(&config->spim);` again right ? While in the original implementation this was only called if the device was ACTIVE.

gmarull replied:
Indeed, this is a PM framework problem because we have poorly defined states and transitions, leading to this sort of issues. In a first place, `pm_device_state_set` needs to be removed from the public API space as we all agree it is not the proper way to handle device PM. Second, most (if not all) devices have this sort of fall-through to handle all "low power states", which is wrong. The possible state transitions should be limited, e.g. as Windows does https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/device-power-states

I have just opened an issue detailing all these problems https://github.com/zephyrproject-rtos/zephyr/issues/37056

#36125 goes further simplifying most of these problems, plus introducing the actions concept, which helps to prevent this sort of problems.

ceolin replied:
Agree, but we should stop break everything thing while we don't have proper solutions. I like the idea of having defined transitions so devices don't need to check it by themselves. But even considering what is done by windows in https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/device-power-states, it is possible to go from D3Hot to D3Cold but it is not defined that all devices should support it.

So lets do an exercise, imagine this devices maps `PM_DEVICE_STATE_SUSPEND` to `D3Hot` and `PM_DEVICE_STATE_OFF` to `D3Cold`. In this scenario the transition between `PM_DEVICE_STATE_SUSPEND` to `PM_DEVICE_STATE_OFF` is allowed from the subsystem perspective, but this driver does not support it. It just change the state if the current state is `ACTIVE`.  With your patch this transition will happen. 

That is wrong ? I don't know, but I don't like change device drivers behavior in this massive changes. I vote to keep their behavior and after this being merged address it in another pr, that is smaller and will get attention from the proper maintainer. We *can't* assume we know everything about all devices.

ceolin replied:
Btw, that is my only concern in this pr right now, everything else is great for me.

gmarull replied:
I think the problem is that most drivers started following the pattern below because there is no clear guidance on what low power states mean as of today, so everything that sounded like "low power" was put into the same basket:

```c
	case PM_DEVICE_STATE_LOW_POWER:
	case PM_DEVICE_STATE_SUSPEND:
	case PM_DEVICE_STATE_OFF:
```

This design pattern has some problems. For example, suspend and off should not perform the same action since one will keep context while the other won't. There are actually many other problems that this PR is not fixing, see e.g.

```c
		if (*state == PM_DEVICE_STATE_ACTIVE) {
			ret = st7735r_exit_sleep(data);
			if (ret < 0) {
				return ret;
			}
			data->pm_state = PM_DEVICE_STATE_ACTIVE;
		} else {
			ret = st7735r_enter_sleep(data);
			if (ret < 0) {
				return ret;
			}
			data->pm_state = PM_DEVICE_STATE_LOW_POWER;
		}
```

In this case things are even worse, since any state `!= PM_DEVICE_STATE_ACTIVE` will call the sleep code and leave the device in `PM_DEVICE_STATE_LOW_POWER`, even if another state was requested.

Note that #36125 RFC fixes most (if not all) these problems in later patches by removing these bad patterns. But still, I think that until #37056 is solved, we won't have a final/proper solution.

gmarull replied:
For reference, the final tentative code proposed in the RFC for the mentioned drivers:

https://github.com/zephyrproject-rtos/zephyr/blob/cebb50a54e375f1e5c7af96ae38d1602a70c2531/drivers/spi/spi_nrfx_spi.c#L280-L303

https://github.com/zephyrproject-rtos/zephyr/blob/cebb50a54e375f1e5c7af96ae38d1602a70c2531/drivers/display/display_st7735r.c#L510-L529

ceolin commented on samples/subsys/pm/device_pm/src/dummy_driver.c:
```diff
@@ -33,7 +33,8 @@ static int dummy_open(const struct device *dev)
 	(void) pm_device_wait(dev, K_FOREVER);
 
-	if (dev->pm->state == PM_DEVICE_STATE_ACTIVE) {
+	pm_device_state_get(dev, &state);
```
`(void)pm_device_state_get(dev, &state);`

gmarull replied:
fixed

gmarull commented on drivers/serial/uart_cc13xx_cc26xx.c:
```diff
@@ -401,69 +398,40 @@ static int postNotifyFxn(unsigned int eventType, uintptr_t eventArg,
 static int uart_cc13xx_cc26xx_pm_control(const struct device *dev,
-					 uint32_t ctrl_command,
-					 enum pm_device_state *state)
+					 enum pm_device_state state)
```
there's an upcoming patch cleaning up driver implementations (https://github.com/zephyrproject-rtos/zephyr/pull/36125/commits/2aecc31fbbed9b98cb2d0af918a66b3d43c000c1), it's out of the scope of this PR.

ceolin replied:
That is not out of this scope, since it is this patch that is causing this function no longer be needed. It is definitely out of scope, but I can live with it being addressed in another pr.

gmarull commented on drivers/serial/uart_npcx.c:
```diff
@@ -467,28 +453,14 @@ static inline int uart_npcx_set_power_state(const struct device *dev,
+static int uart_npcx_pm_control(const struct device *dev,
+				enum pm_device_state state)
+{
+	return uart_npcx_set_power_state(dev, state);
```
ref. #36125

gmarull commented on drivers/spi/spi_nrfx_spim.c:
```diff
@@ -323,52 +320,32 @@ static int init_spim(const struct device *dev)
+	case PM_DEVICE_STATE_LOW_POWER:
+	case PM_DEVICE_STATE_SUSPEND:
+	case PM_DEVICE_STATE_OFF:
+		nrfx_spim_uninit(&config->spim);
```
https://github.com/zephyrproject-rtos/zephyr/pull/36759#discussion_r672528317

ceolin replied:
https://github.com/zephyrproject-rtos/zephyr/pull/36759#discussion_r672548573

gmarull commented on subsys/pm/device.c:
```diff
@@ -129,23 +129,58 @@ const char *pm_device_state_str(enum pm_device_state state)
+	case PM_DEVICE_STATE_FORCE_SUSPEND:
+		__fallthrough;
+	case PM_DEVICE_STATE_LOW_POWER:
+		__fallthrough;
```
they all end up running the same check, since they don't have transitional states:
```
		if (dev->pm->state == state) {
			return -EALREADY;
		}
```

ceolin replied:
gotcha.

ceolin replied:
that is correct.

gmarull commented on drivers/sensor/bmp388/bmp388_trigger.c:
```diff
@@ -92,7 +92,10 @@ int bmp388_trigger_set(
-	if (data->device_power_state != PM_DEVICE_STATE_ACTIVE) {
+	enum pm_device_state state;
+
+	pm_device_state_get(dev, &state);
```
fixed

nashif reviewed (changes requested):
Changing the control API requires further discussion, the cleanup in general is ok, but it does not provide an alternate long-term solution.

gmarull commented on drivers/display/display_st7735r.c:
```diff
@@ -519,37 +512,22 @@ static int st7735r_enter_sleep(struct st7735r_data *data)
 
-static int st7735r_pm_control(const struct device *dev, uint32_t ctrl_command,
-			      enum pm_device_state *state)
+static int st7735r_pm_control(const struct device *dev,
```
This is why actions (or call them commands) are introduced in the #36125 RFC.

gmarull replied:
In the end, all these changes simplify device PM and at the same time align a bit more with what Linux does. I think these changes do not fix all the problems of device PM, but I think they move to the right direction. If one looks at device PM today, it's really in a sad state (very few drivers implement it and most have bugs). #36125 or #37056 provide enough evidence to prove that. I haven't seen any alternatives or roadmap to improve the large amount of PM issues we have today.

nashif replied:
> This is why actions (or call them commands) are introduced in the #36125 RFC.

ok, that was exactly on my mind and makes sense. You need better isolate changes in separate PRs. This PR on its own does not make sense without the move to actions. The other Pr is full of conflicts and I am not sure anymore which is which, every PR starts with "simplify"... :)

gmarull replied:
Just rebased #36125 so that it's easier to browse.

gmarull commented:
Before doing any more changes, I'd like to have a clear roadmap on where we want to go, otherwise every PR will take ages to merge and as a result we'll keep having a device PM subsys that is far from solving the PM problem. I think that if the goal is clear, and the current PM subsys status is well understood/recognized, things will be easier. I have the feeling that device PM in particular is _sold_ as ready when it is clearly not, and, as far as I know, I'm not the first person to mention that.

Let's take a couple of examples to illustrate my argument:

- The `ctrl_command` argument was, in theory, added to support arbitrary commands. The reality, however, is that it has only been used to set/get the device state (with a command name `PM_DEVICE_STATE_GET/SET`, same prefix as actual device states). I think everybody agrees that keeping local states is, in general, a bad idea. Removing the "get", we're left with a single command to set the state. So in reality, the callback is de-facto being used with an _action_, that is, we combine a "set" command with a "state" argument. An example to make it clear: `"set + active" == "resume"`. The `void *context` was replaced in the last release in favor of a `uint32_t *state` and now with `enum pm_device_state *state`, meaning a generic command can hardly be implemented now. So the reason to keep `ctrl_command` doesn't hold anymore. What this PR does in the end is to continue the same design pattern (use states as actions), just removing a redundancy (the ctrl_command), so I don't see what is wrong with this approach as an intermediate step to a proper solution as suggested in #36125.
- The runtime API is another example of something that has existed, but nobody (except a single driver) uses it. It lacks proper documentation and usage guidelines, it doesn't handle device dependencies, etc. Let's talk about a problem I have already mentioned with this API. It basically puts devices into `ACTIVE` or `SUSPEND` states. But what about `LOW_POWER`? Let's see: according to the specs, `LOW_POWER` should be used when context is not lost. In many devices this holds for low power states, so a driver for such device will pick `LOW_POWER`. But then, how can runtime PM be implemented? What happened in reality: most drivers handle all the "low power states" the same way, which is wrong and leads to inconsistent behavior (it's incompatible to implement low power/suspend/off the same way). This is a common pattern in the PM subsystem: there are many APIs that have been created without really analyzing if they can solve a real problem.

My feeling is that the PM subsystem was designed in a rush, with a quick implementation that likely solved immediate problems, but it is now showing its limitations and requires some fundamental changes. We're not talking about some fixes, API extensions, etc. but a refactor/re-write in some areas. This is difficult, painful and likely means a period of some instability, but it is sometimes necessary to make progress. And let's be honest: we'll not either get it 100% right this time.

ceolin commented:
The cleanup is fine, that is definitely not the final version of this control API, but lets move forward.

gmarull committed:
pm: remove incorrect usages of force suspend

Some devices are using PM_DEVICE_STATE_FORCE_SUSPEND as a sort of low
power state, something that is not correct. In fact, this state is not
an actual state and will be eventually moved, if found necessary, to an
action or command.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: converge to suspend state for low power modes

The difference between low power and suspend states is a thin blur line
that is is not clear and most drivers have used indistinctly. This patch
converges to the usage of the suspend state for low power, since
contrary to the low power state, it is used by both system and runtime
device PM. The low power state is still kept, but its future is unclear
and needs some discussion.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented:
@ceolin @nashif I've added all the changes up to initial support of actions.

gmarull committed:
pm: device: handle devices not supporting requested state

If a device does not support, for example, the suspend action the
subsystem should just ignore the device.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
pm: device: add compatibility for force suspend and low power

Add support for FORCE_SUSPEND and LOW_POWER actions. Even though these
actions have no clear meaning, they are added for compatibility until
their associated states are discussed. Their usage in new code should be
discouraged until the associated states are clarified.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>
